csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll

```diff
@@ -3204,6 +3205,275 @@ module ControlFlow {
+           */
+          private predicate firstDefCondition(ConditionBlock cb) {
+            exists(LocalScopeVariableRead read |
+              this.defConditionReachableFromRead(cb, read) |
+              firstReadSameVar(def, read)
+            )
+          }
+
+          override predicate correlatesConditions(ConditionBlock cb1, ConditionBlock cb2, boolean inverted) {
+            this.firstDefCondition(cb1) and
+            exists(LocalScopeVariableRead read1, LocalScopeVariableRead read2 |
+              read1 = cb1.getLastElement() and
+              adjacentReadPairSameVar+(read1, read2) and
+              read2 = cb2.getLastElement() and
+              inverted = false
+            )
+          }
+
+          override Callable getEnclosingCallable() {
+            result = def.getVariable().getCallable()
+          }
+
+          override string toString() {
+            result = def.getVariable().toString()
+          }
+
+          override Location getLocation() {
+            result = def.getLocation()
+          }
+        }
+
+        /**
+         * A split for elements that can reach a condition where this split determines
+         * the Boolean value that the condition evaluates to. For example, in
+         *
+         * ```
+         * if (b)
+         *     Console.WriteLine("b is true");
+         * if (!b)
+         *     Console.WriteLine("b is false");
+         * ```
+         *
+         * all control flow nodes on line 2 and line 3 have two splits: one representing
+         * that the condition on line 1 took the `true` branch, and one representing that
+         * the condition on line 1 took the `false` branch.
+         */
+        class BooleanSplitImpl extends SplitImpl, TBooleanSplit {
+          private BooleanSplitSubKind kind;
+          private boolean branch;
+
+          BooleanSplitImpl() {
+            this = TBooleanSplit(kind, branch)
+          }
+
+          /** Gets the kind of this Boolean split. */
+          BooleanSplitSubKind getSubKind() { result = kind }
+
+          /** Gets the branch taken in this split. */
+          boolean getBranch() { result = branch }
+
+          override string toString() {
+            exists(int line |
+              line = kind.getLocation().getStartLine() and
+              result = kind.toString() + " (line " + line + "): " + branch.toString()
+            )
+          }
+        }
+
+        private class BooleanSplitKind extends SplitKind, TBooleanSplitKind {
+          private BooleanSplitSubKind kind;
+
+          BooleanSplitKind() {
+            this = TBooleanSplitKind(kind)
+          }
+
+          /** Gets the sub kind of this Boolean split kind. */
+          BooleanSplitSubKind getSubKind() { result = kind }
+
+          override int getListOrder() {
+            exists(Callable c, int r |
+              c = kind.getEnclosingCallable() |
+              result = r + 1 and // start the ordering from 2
```

Comment:
LGTM - however I anticipate this conflicting with other list kinds - unless they are numbered from 100 or something?

Reply:
Before this PR, we only had orders `0` and `1`, for `finally` splits and exception handler splits, resp. These are easy, because the orders are static. For Boolean splits, we have one kind per SSA variable that we want to split on, so we need to generate the orders dynamically, as above. If/when we are to add new kinds later, we can either add them as statically as `2` (and change the code above), or as `max(BooleanSplitKind k | k.getListOrder()) + 1`.